### ISSUE TITLE
R CMD check using unexpected collate settings

### ISSUE THREAD

**Issue by bersbersbers** [r-lib/devtools#2377]
We have noticed in https://github.com/rstudio/shinytest/issues/409 that when `devtools::check()` calls `shinytest`, keys are sorted in different order on some Linux systems. It seems this has something to do with collate settings, but we are unable to see where these are set and how we can unset them.

Here's an example package:

```
$ tree
.
├── DESCRIPTION
├── inst
│   └── shiny
│       ├── app.R
│       └── tests
│           └── shinytest
│               ├── repro_test-expected
│               │   └── repro_test.json
│               └── repro_test.R
└── R
    └── repro.R

6 directories, 5 files

$ cat DESCRIPTION
Type: Package
Package: repro
Title: Repro
Description: This is a description.
Author: Author
Maintainer: Maintainer <main@tain.er>
License: GPL-3
Version: 0.0.1
Depends: shiny
Suggests: shinytest
Encoding: UTF-8
RoxygenNote: 7.1.2

$ cat inst/shiny/app.R
Sys.setlocale("LC_ALL", "en_US.UTF-8")
shinyApp(repro:::repro_ui(), repro:::repro_server)

$ cat inst/shiny/tests/shinytest/repro_test.R
app <- ShinyDriver$new("../../", seed = 0)
app$snapshotInit("repro_test", screenshot = FALSE)
app$snapshot(filename = "repro_test.json")
# Sys.sleep(120)

$ cat inst/shiny/tests/shinytest/repro_test-expected/repro_test.json 
{
  "input": {
    "X_b": "X_b",
    "Xa": "Xa"
  },
  "output": {

  },
  "export": {

  }
}

$ cat R/repro.R 
repro_server <- function(input, output, session) {}

#' @import shiny
repro_ui <- function() {
  fluidPage(
    selectInput("Xa", "Xa", choices = "Xa"),
    selectInput("X_b", "X_b", choices = "X_b"),
  )
}

#' Run tests
#'
#' @examples
#' run_tests()
#' @export
run_tests <- function() {
  shinytest::testApp(system.file("shiny", package = "repro"))
}

```

From this filestructure, run 

```bash
R -e "shinytest::testApp('inst/shiny')"
```
to verify that the test passes. 

(You can also do
```
R -e "devtools::install(); repro::run_tests()"
```
with the same result: pass.)

Then, run
```bash
R -e "devtools::check()"
```
Then, the same test (run as an example) fails.

To see why, uncomment `Sys.sleep(120)` above and re-run `check()`. While `shinytest` is sleeping, check the output using `find /tmp/ | grep current/repro_test.json`:

```json
{
  "input": {
    "Xa": "Xa",
    "X_b": "X_b"
  },
  "output": {

  },
  "export": {

  }
}
```

Note the sort order is different.

On my system, it is expected that collate settings influence the sort order:

```
 version  R version 4.1.1 (2021-08-10)
 os       openSUSE Leap 15.2          
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 

```

```bash
$ printf "Xa\nX_b" | LC_COLLATE=en_US sort
Xa
X_b
$ printf "Xa\nX_b" | LC_COLLATE=C sort
X_b
Xa
```

However, we are unable to reproduce this within R:
```bash
$ LC_COLLATE=en_US R -e "sort(c('Xa', 'X_b'))"
> sort(c('Xa', 'X_b'))
[1] "X_b" "Xa" 
> 
$ LC_COLLATE=C R -e "sort(c('Xa', 'X_b'))"
> sort(c('Xa', 'X_b'))
[1] "X_b" "Xa" 
> 
```
or
```r
> x <- c('Xa', 'X_b')
> withr::with_collate("C", sort(x))
[1] "X_b" "Xa" 
> withr::with_collate("en_US.UTF-8", sort(x))
[1] "X_b" "Xa" 
> 
```

So the issue is, why is the sort order changing inside `devtools::check()`? And if this is through collate settings, why are we unable to emulate the same effect?

It seems to have something to do with `shiny`/`shinytest`: if I replace `run_tests` simply by

```r
run_tests <- function() {
  stopifnot(all(order(c("X_b", "Xa")) == c(1, 2)))
}
```
then the example code runs fine.

**Comment by bersbersbers** [comment]
The reprex above was apparently created while I had https://github.com/rstudio/shiny/pull/3515 installed. This PR did not solve the issue for me. However, it may be that this reprex only reproduces the issue with this PR - at least, I have trouble reproducing this using this reprex after reinstalling release `shiny`. I amended it with `Sys.setlocale("LC_ALL", "en_US.UTF-8")` which reproduces the issue.

**Comment by jimhester** [comment]
The collation order is part of the locale database on the operating system. Yes the collation differs on macOS vs on most linux systems for the en_US locale, on macOS the sorting is case insensitive, and basically just does a byte comparison. e.g. https://blog.zhimingwang.org/macos-lc_collate-hunt

I don't think this has anything to do with devtools.

**Comment by bersbersbers** [comment]
@jimhester agreed - the issue may be less severe on MacOS. But just to most clear: the issue is that on the same system, a test passes when called directly, and fails when run as an example through `devtools::check()`.

**Comment by jimhester** [comment]
And what happens with `R CMD check`?

**Comment by bersbersbers** [comment]
First, following https://github.com/r-lib/devtools/issues/2377#issuecomment-930529248, I have been able to reduce the reprex to just this:

```
$ tree
.
├── DESCRIPTION
└── R
    └── repro.R

$ cat DESCRIPTION 
Type: Package
Package: repro
Title: Repro
Description: This is a description.
Author: Author
Maintainer: Maintainer <main@tain.er>
License: GPL-3
Version: 0.0.1
Encoding: UTF-8
RoxygenNote: 7.1.2

$ cat R/repro.R 
#' Run tests
#'
#' @examples
#' run_tests()
#' @export
run_tests <- function() {
  Sys.setlocale("LC_ALL", "en_US.UTF-8")
  stopifnot(all(order(c("X_b", "Xa")) == c(1, 2)))
}
```

No more dependence on `shiny`/`shinytest`.

> And what happens with `R CMD check`?

A few outputs:

```
$ R -e 'Sys.setlocale("LC_ALL", "en_US.UTF-8"); stopifnot(all(order(c("X_b", "Xa")) == c(1, 2)))'
--> PASS

$ R -e "devtools::install(); repro::run_tests()"
--> PASS

$ R CMD build . && R CMD check *.gz
--> (I don't see it running examples - it says "checking examples ... NONE" - is that PASS?)

$ R CMD check .
--> FAIL: Error in run_tests() : all(order(c("X_b", "Xa")) == c(1, 2)) is not TRUE

$ R -e "devtools::check()"
--> FAIL:   Error in run_tests() : all(order(c("X_b", "Xa")) == c(1, 2)) is not TRUE
```

So, yes - devtools does not seem to be involved. Where to turn next? ;)

**Comment by bersbersbers** [comment]
For the record, the minimum example I could come up with:

```r
$ tree
.
├── DESCRIPTION
└── R
    └── repro.R

$ cat DESCRIPTION 
Type: Package
Package: repro
Title: Repro
Description: This is a description.
Author: Author
Maintainer: Maintainer <main@tain.er>
License: GPL-3
Version: 0.0.1
Encoding: UTF-8
RoxygenNote: 7.1.2

$ cat R/repro.R 
#' Run tests
#'
#' @examples
#' run_tests()
#' @export
run_tests <- function() {
  Sys.setlocale("LC_ALL", "en_US.UTF-8")
  if (!all(order(c("_a", "a")) == c(1, 2))) stop("FAIL")
  message("PASS")
}
```

Then run
```shell
R -e "devtools::document()" # prepare
R -e "devtools::install(); repro::run_tests()" # PASS
R CMD check . # FAIL
```

**Comment by bersbersbers** [comment]
By the way, this only happens with `order(..., method="shell")`, not with `order(..., method="radix")`.

**Comment by bersbersbers** [comment]
The core issue seems to be independent of `R CMD check` as well.

In fact, the `R CMD check` issue has been discussed before: https://stackoverflow.com/q/42272119/, and I have confirmed that the test, when run through `check` has an additional `LC_COLLATE=C` environment variable set. (I have not been able to find where that happens in the code, but there is a bunch of other places in the code where this is done.)

But there is more to learn. For that, it is relevant that my system has `LANG=en_US.UTF-8` set, and no `LC_*` at all.

Now, `LANG` explains why the default `printf "_a\na" | sort` gives `a` first. Let's call this 'EN' order.

Namely, `printf "_a\na" | LANG= sort` gives `_a` first.

Similarly, `printf "_a\na" | LC_COLLATE=C sort` gives `_a` first. So let's call this `C` order.

And then let's look at what R does:

1. `R -e 'sort(c("_a", "a"), method="shell")'`
[1] "_a" "a" 
This ignores `LANG` and uses the `C` order on my system. (I can live with that.) Same with `LANG=` and `LC_COLLATE=C`:
`LANG= R -e 'sort(c("_a", "a"), method="shell")'`
`LC_COLLATE=C R -e 'sort(c("_a", "a"), method="shell")'`
 Alright.

2. `R -e 'Sys.setlocale("LC_COLLATE", "en_US.UTF-8"); sort(c("_a", "a"), method="shell")'`
[1] "_a" "a" 
Note this is still `C` order: setting the `en_US ` locale is ineffective. I consider this a bug. This is true even in the absence of `LANG`:
`LANG= R -e 'Sys.setlocale("LC_COLLATE", "en_US.UTF-8"); sort(c("_a", "a"), method="shell")'`
[1] "_a" "a" 

3. `LC_COLLATE=C R -e 'Sys.setlocale("LC_COLLATE", "en_US.UTF-8"); sort(c("_a", "a"), method="shell")'`
[1] "a"  "_a"
This is the crazy part! It is only by setting `LC_COLLATE=C` outside of R that setting `LC_COLLATE` to `en_US` becomes effective inside R, and we obtain `EN` order for the first time.
By the way, this can *not* be emulated inside R:
`R -e 'Sys.setlocale("LC_COLLATE", "C"); Sys.setlocale("LC_COLLATE", "en_US.UTF-8"); sort(c("_a", "a"), method="shell")'`
[1] "_a" "a"

So, TLDR:
- There is to sort methods that can behave differently.
- R is able to reproduce the behavior of command-line `sort` in all aspects I looked at. It is not the case that R `sort` is somehow more stable by design.
- `Sys.setlocale("LC_COLLATE", ...)` can be ineffective if `LC_COLLATE` has *not* been set before.
- The code as run by `R CMD check` is actually the code that is "correct" (in line with the chosen locale in the line immediately before); and the code as run standalone is the incorrect one, because in it, for some reason, changing the locale in the line before is partially ineffective.

**Comment by bersbersbers** [comment]
Final message for tonight (this is getting crazier and crazier): when I write `C` order and "partially" ineffective, I mean with respect to underscores. In total, there is more combinations:

The "default" is one `C` order (but not the one I get from `LC_COLLATE=C sort`):
```
$ R -e 'sort(c("_a", "a", "A", "b"), method="shell")'
$ LANG= R -e 'sort(c("_a", "a", "A", "b"), method="shell")'
$ R -e 'Sys.setlocale("LC_COLLATE", "en_US.UTF-8"); sort(c("_a", "a", "A", "b"), method="shell")'
[1] "_a" "a"  "A"  "b" 
```

This is another `C` order (actually the one `LC_COLLATE=C sort` gives me):
```
$ LC_COLLATE=C R -e 'sort(c("_a", "a", "A", "b"), method="shell")'
$ R -e 'Sys.setlocale("LC_COLLATE", "C"); sort(c("_a", "a", "A", "b"), method="shell")'
[1] "A"  "_a" "a"  "b" 
```

This is `EN` order:
```
LC_COLLATE=C R -e 'Sys.setlocale("LC_COLLATE", "en_US.UTF-8"); sort(c("_a", "a", "A", "b"), method="shell")'
```

**Comment by bersbersbers** [comment]
For the record, Windows gives me

```
> Sys.getlocale()
[1] "LC_COLLATE=English_United States.1252;LC_CTYPE=English_United States.1252;LC_MONETARY=English_United States.1252;LC_NUMERIC=C;LC_TIME=English_United States.1252"
> sort(c("_a", "a", "A", "b"))
[1] "_a" "a"  "A"  "b"
```

in line with
```
>more.com x | sort.exe
_a
A
a
b
```

and which equals what I see get on Linux R by default - but what I cannot reproduce using Linux `sort`.

**Comment by bersbersbers** [comment]
I am losing track of what is the "expected" output, but at least I see now that `Sys.setlocale` does not set all aspects of the locale relevant for `sort` - we need `Sys.setenv` as well.

Consider this code:
```r
env <- \() Sys.setenv(LC_COLLATE = "en_US.UTF-8")
loc <- \() Sys.setlocale("LC_COLLATE", "en_US.UTF-8")
run <- \() sort(c("_a", "a", "b", "A", "B"), method = "shell")
```

and these runs:
```shell
$ env -i R -q -e "source('debug.R'); run()"
[1] "A"  "B"  "_a" "a"  "b" 

$ env -i R -q -e "source('debug.R'); env(); run()"
[1] "A"  "B"  "_a" "a"  "b" 

$ env -i R -q -e "source('debug.R'); loc(); run()"
[1] "_a" "a"  "A"  "b"  "B" 

$ env -i R -q -e "source('debug.R'); env(); loc(); run()"
[1] "_a" "a"  "A"  "b"  "B" 
```
Above: if `LC_COLLATE` is not set externally, `env()` does not change anything.

```shell
$ env -i LANG=en_US.UTF-8 LC_COLLATE=C R -q -e "source('debug.R'); run()"
[1] "A"  "B"  "_a" "a"  "b" 

$ env -i LANG=en_US.UTF-8 LC_COLLATE=C R -q -e "source('debug.R'); env(); run()"
[1] "A"  "B"  "_a" "a"  "b" 

$ env -i LANG=en_US.UTF-8 LC_COLLATE=C R -q -e "source('debug.R'); loc(); run()"
[1] "a"  "_a" "A"  "b"  "B" 

$ env -i LANG=en_US.UTF-8 LC_COLLATE=C R -q -e "source('debug.R'); env(); loc(); run()"
[1] "_a" "a"  "A"  "b"  "B" 
```
With `LC_COLLATE=C` set externally, we need to overwrite it with `env()` locally, even if we already set the locale locally with `loc()`.

**Comment by lionel-** [comment]
The explanation is in https://github.com/wch/r-source/blob/0d8003afefdfbcfe95db51d88c6b6ab8925c4a05/src/main/util.c#L2458-L2468.

R supports setting `LC_COLLATE` to "C" via an envvar. When that is the case, the envvar has precedence over the currently set locale. The different orderings that you see with different order methods are because only sort methods using `Scollate()` are affected by this rule.

Fixed on the withr side in https://github.com/r-lib/withr/commit/e3dd4b08935c39620d7c56539794773e2d07a1c5. This means that code ran within testthat blocks will no longer be affected by this, which should solve your `devtools::check()` issues.

**Comment by bersbersbers** [comment]
For the record, this does not solve the inconsistencies seens in https://github.com/r-lib/devtools/issues/2377#issuecomment-931026796 (but these do not not seen to be directly related to `devtools` or `withr`, either, so it's probably fine to close these here). If someone has an idea where to repost this, I'll be happy to do so.

**Comment by lionel-** [comment]
This would be R itself. But I don't think they'll take the risk to change this behaviour at this point.

In any case, `withr::local_collate()` is automatically called in `test_that()` blocks, so that should solve the issue within testthat tests.
